### DISCUSSION TITLE
Document View for JSON

### DISCUSSION THREAD

**Discussion by falkreon** [falkreon/Jankson#53] [Design Goals]
When I say "Document" I don't mean `org.w3c.dom.Document`, but it's on my mind. Preserving ordering and comments are a pillar of this library and honestly some of the edge-case behavior is still unacceptable.

Currently, comments are attributed to a node in the node-list (usually a key-value pair). But what happens in this case?
```json5
{
  foo: //this is a key
  10 //this is a value
}
```

The best thing we could do after a roundtrip right now is this
```json5
{
  /* this is a key
  this is a value */
  foo: 10
}
```

And I'm really not sure we're even doing that. Worse, what do we do here?
```json5
/* this object blah blah blah
...
blah. */
{
  foo: 10
}
```

So every object needs to also carry around a "top-level comment" field?

What I want to do, and this is going to be a big project, is to create a structural view of the file. That is, an object is a sub-file and contains individual nodes for keys, objects, multiline comments, end-of-line comments, etc, and can be used to more or less recreate the original file, *without* moving or merging comments. We will fix indentation and any errors we find along the way, and in some ways regularize the file. This is important for jankson's usefulness as a preprocessor. But this would unhook the comments from the key-value pairs and give us a foothold up to handling all the edge cases.

Does such an object continue to support random access and structural modifications? Do we replace the implementation for JsonObject, or hold these two views into the file as distinct - one document view and one semantic view, and let the user pick which one they want to work with for any given task?

This is a hard design problem, and I'm looking for input on how we could get it done cleanly and clearly.

**Comment by Niko-sk2x** [comment]
I think the most reasonable way to deal with it would be for comments to be their own type of entry inside any json object or array, and also have a top level json "container" for the entire file. That would naturally preserve the order and allow to keep any extra information.

It would mean they are no longer associated with any given elements, but there could be a way to access an element with comment context around it, and then it would be on the user of that library to decide how to interpret it when manipulating the data.

In json object it wouldn't be associated with any key, but could be accessible in a collection of all entries.

Edit: I didn't think about comment being in the middle of a key-value pair. At that point it would also require putting an optional comment field in each a json object entry.

Alternatively, comments could be be considered more as separate things in each json object and array, with information about which elements are before and after it (possibly including information whether it's on the same line or a different line?). 

Moving, removing and adding elements would need an option to specify how to deal with comments around it.

But at least for my use case, I think the most natural way to represent it would be for json objects to just expose a collection of some sort of "json object components" which could be: `comment`, `key`, `value` (possibly as a linked list type of structure, where each element would just have a reference to previous and next). And also allow constructing json objects from such elements. But my use case is quite unusual and probably wouldn't work for everyone.

**Reply by falkreon** [reply]
this is kind of where I'm going with this. a "document view" has a document root, and a hierarchial collection of objects like key, value, comment, which can be combined to form objects and arrays. I envision something like `List<String> getRelatedComments(key)`, with a well specified heuristic of where such comments could come from (following more or less the current attribution rules). but you'd need to be able to shift between the structural view and the semantic one depending on your needs.

**Reply by Niko-sk2x** [reply]
One thing I would definitely need to be able to do is add/move/remove/generate comments in existing JSON so any solution that can't do that would be a big issue for me. `getRelatedComments` that just returns a `List<String>` doesn't look particularly useful unless it would be possible to specify the heuristic. It wouldn't be of any use for preserving comments when manipulating the JSON in any way.

A lot also depend on whether the target is to preserve information about the exact type of comment (`//` vs `/* */`, newlines, dealing with indentation).

If this information should be preserved, the document view would probably also have to preserve line breaks in general. Especially difficult case seems to be dealing with `/* comment1 */ "key" /* comment2 */ : /* comment3 */ "value" /* comment4 */, /* comment5 */ "nextKey" ...`. In that case the placement of key/value separator becomes important.
For this it should be possible for comments to be attributed to "key document entry" and "value document entry" with "before" and "after", but that still doesn't deal with line breaks (but then each comment entry could specify whether it has a line break before/after the comment, but that could also result in ambiguity with multiple comments following each other. Another option would be to just not store information about line breaks). And there would have to be separate "end of line" comment, that would probably be best associated with the last key on that line, or the previous. And aside of that there would still have to be separate "comment" entry for comments not associated with anything. Whole line `//` style comments just before a key would be associated with the key right after them.

So to show an example (using `#` as comment on how it would be represented, note that there is a difference between endOfLine "before"/"after" comment entry and endOfLine comment entry associated with an element) :

// EDIT: this is to be improved, I will try to cover every case and think about it in more detail
```json5
// comment # JsonDocumentComment at the beginning of document root, no attribution, isEndOfLine=true
{ # this begins JsonDocumentObject inside the document root
  /* comment */  # "before" comment on key, isEndOfLine=false
  // comment # another "before" comment on key, isEndOfLine=true
  /* comment1 */ "key" /* comment2 */ // comment3 # comment1 is yet another "before" comment on key, comment2 is "after" comment on key, comment3 is end of line comment on key
    /* comment4 */ : /* comment5 */ // comment6 # comment4 is another "after" comment on key, comment5 is "before" comment on value, comment6 is "before" endOfLine comment on value
    /* comment7 */ "value" /* comment8 */, // comment9 # comment7 is another "before" comment on value, comment8 is an "after" comment on value, comment9 is an end of line comment associated with value
  key2: "value2" // comment10 # comment10 is an "after" endOfLine comment on value2
  // comment11 # comment11 is another "after" endOfLine comment on value2
  , // comment12 # comment12 is separate comment entry
    /* comment13 */ # so is comment13, but not of end of line type. line breaks would change comment attribution
  key3: "value3" // comment14  # comment14 is end of line comment for key3 
  // comment15 # separate comment entry
} // ends an object
// comment # separate comment entry in document root
```


And the existing attribution style comments should probably stay as an option since not everyone needs or wants to deal with comments on such detailed level.

Code dealing with such comments (here just printing them out) would look something like this:
```java
for (JsonObjectDocumentEntry entry : documentObject.entries()) {
  if (entry.isComment()) {
    System.out.println("Comment: " + entry);
    System.out.println("\tisEndOfLine: " + entry.asComment().isEndOfLine());
  }
  else if (entry.isJsonComponent()) {
    System.out.println("key: " + entry.asJsonComponent().getString());
    for (JsonDocumentCommentEntry c : entry.asJsonComponent().commentsBefore()) {
      System.out.println("\tcommentBefore: " + c);
    }
    for (JsonDocumentCommentEntry c : entry.asJsonComponent().commentsAfter()) {
      System.out.println("\tcommentsAfter: " + c);
    }
    if (entry.asJsonComponent().endLineComment() != null) {
      System.out.println("\tendLineComment: " + entry.asJsonComponent().endLineComment());
    }
 }
```

There could also be a view as a stream/list of json elements, something like what gson has with `JsonReader` (except with actual object hierarchy instead of just a direct stream of tokens). 

One particular use case I have that currently works well enough only because the existing json files where this needs to be done the most aren't allowed to have comments (as they were originally meant for parsing with GSON) is transforming json like this (this exact one is guaranteed to not have comments, but I'm planning to add options that will make it necessary to actually deal with cases that can in future versions):
```json
// comment1
"coalOreSpawnSize": value,
// comment2
"coalOreSpawnTries": value,
"coalOreSpawnProbability": value,
"coalOreSpawnMinHeight": value,
"coalOreSpawnMaxHeight": value,
"ironOreSpawnSize": value,
"ironOreeSpawnTries": value,
// comment3
"ironOreSpawnProbability": value,
"ironOreSpawnMinHeight": value,
"ironOreSpawnMaxHeight": value
```
into
```json5
"standardOres": [
  {
    blockstate: { Name: "minecraft:coal_ore" },
    // comment1
    spawnSize: value,
    // comment2
    spawnTries: value,
    spawnProbability: value,
    minHeight: value,
    maxHeight: value,
    biomes: null,
    genInBlockstates: null
  },
  {
    blockstate: { Name: "minecraft:iron_ore" },
    spawnSize: value,
    spawnTries: value,
    // comment3
    spawnProbability: value,
    minHeight: value,
    maxHeight: value,
    biomes: null,
    genInBlockstates: null
  }
]
```

The above suggested API would make it possible for me to properly move the comments around while converting to new data format.

**Reply by falkreon** [reply]
Agreed, there should be a hot-path for comment attribution. I'm not really committed to preserving the *kind* of comment, I think that falls under the category of formatting-and-cleanup, and how we handle the comments can be good feedback of how the comments were attributed.

In your example (stripped of comment comments)
```json5
{
  /* comment-1 */ 
  // comment0
  /* comment1 */ "key" /* comment2 */ // comment3
    /* comment4 */ : /* comment5 */ // comment6
    /* comment7 */ "value" /* comment8 */, // comment9
  key2: "value2" // comment10
}
```
I don't see comment-1, comment0, or comment1 as distinct, and I'm hard-pressed to come up with clean rules for why that should be the case. I don't like relying on line breaks except to delimit line-end comments. One possibility is to say, if there are multiple comments "before" a key-value-pair or element, then only the closest comment prevails, *with the exception that "bare" line-end comments are merged* (bare as in, nothing precedes them on the line except whitespace). So what I'm proposing is:

```json5
/* a
*/
// b
// c
// d
{
  //e
  /* f */
  key1: "value1"
}
```

`a` is considered to be a document preamble because it is not the first comment before the root object
`b\nc\nd` is the commentBefore the root object
`e` is an object preamble
`f` is the commentBefore key1

comments stuffed into the cracks between the key and value would mostly be retrievable with the document api, whose view would be more like

```json5
documentRoot {
  comment
  lineEndComment
  lineEndComment
  lineEndComment
  objectElement {
    lineEndComment
    comment
    key //note that the separator is not a node; comments before the separator may be forced after the separator
    primitiveElement
  }
}
```

**Reply by falkreon** [reply]
Side note I like the phrasing `asComment`, `asJsonElement`, etc., we need to keep that.

**Reply by Niko-sk2x** [reply]
The reason I thought line breaks matter is that this:
```json5
{
  "key": "value", // comment
  "key2": "valye2"
}
```
is in user's view very different from
```json5
{
  "key": "value", 
  // comment
  "key2": "valye2"
}
```

In the first case, the comment obviously belongs to the first key-value pair (which is why I came up with the distinction for end of line comments)
In the second case it seems like it should be attributed to key2.

Dealing with them as completely separate without attribution and not preserving type solves that. Then you would only have separate `after key` and `before value` and `after value` distinction for placement around a separator. And not keeping the type of comment would be fine, I just wasn't sure if you wanted to do that.

Then we would have the following:
```json5
{
  // comment1
  "key" /* comment2 */: /* comment3 */ "value" /* comment4 */, // comment5
  // comment6
}
```
`comment1` and `comment5\ncomment6` would be standalone comments. `comment2` would be `jsonObjectKey.commentAfter()`, `comment3` would be `jsonObjectValue.commentBefore()` and `comment4` would be `jsonObjectValue.commentAfter()`. 

Object representation could compared to your above example look like this:
```
documentRoot {
  objectElement {
    standaloneComment: "comment1"
    objectEntry {
      afterKeyComment: "comment2"
      key: "key"
      beforeValueComment: "comment3"
      afterValueComment: "comment4"
      value: primitiveElement { "value" }
    }
    standaloneComment: "comment5\ncomment6"
  }
}
```
(2 comments could be kept as separate instead of with newlines)
And it could get reformatted as:
```json5
{
  /* comment1 */
  "key" /* comment2 */: /* comment3 */ "value" /* comment4 */, 
  /* comment5
  comment6 */
}
```
or alternatively with an option to prefer `//` style comments when they are on their own line:
```json5
{
  // comment1
  "key" /* comment2 */: /* comment3 */ "value" /* comment4 */, 
  // comment5
  // comment6
}
```
Or possibly with an option to force `//` comments:
```json5
{
  // comment1
  "key" // comment2
  : // comment3
  "value" // comment4
  , 
  // comment5
  // comment6
}
```

In json arrays, the only possible distinction may have to be "after value" comment to distibguish between
```json5
[
  "value", // comment
  "value"
]
```
and
```json5
[
  "value" /* comment */,
  "value"
]
```

**Reply by falkreon** [reply]
```json5
[ "value" /*comment*/ "value" ]
```
this is a legal construct in jankson. I don't think commentAfter is particularly useful, but if it exists, then it's possible for the commentAfter to be the commentBefore of the next element.

**Reply by falkreon** [reply]
As I get further into the nuts and bolts of implementing such a thing, it seems like we want to make a "KeyValuePair" intermediate entry which itself holds a list of entries associated with it. This would help clean up the semantics for what `commentBefore` and `commentAfter` might eventually mean:

```json5
//things
/* stuff */
{
    //thing
    //other thing
    "status": /* foo */
    [ 4 ] // number chosen by a fair diceroll; guaranteed to be random
    //bar
} // a
//b
```

Possible object tree:
```json5
documentRoot {
    commentEntry "things" // document preamble
    objectEntry { //not synthetically generated for "bare root objects". if the root was bare, it's bare in the doc.
        commentBefore: commentEntry "stuff" //multiple consecutive single-line comments would be folded into a multiline here
        commentAfter: commentEntry "a" //only because it's a line-end comment on the same line
        ---
        keyValueEntry {
            commentBefore: commentEntry "thing\nother thing" //folded
            commentAfter: commentEntry "number chosen by a fair diceroll; guaranteed to be random"
            ---
            primitiveValueEntry "status" //key
            commentEntry "foo" //not commentBefore arrayEntry because it's not "bare"
            arrayEntry {
                commentBefore: null
                commentAfter: null //consumed by keyValueEntry? same as keyValueEntry? TBD
                ---
                primitiveValueEntry 4
            }
        }
        commentEntry "bar"
    }
    commentEntry "b" //Not on the same line as the closing brace, so this is a document footer, along with all following comments.
}
```

When reserializing part of a document, the commentBefore is serialized, then the object itself (including any sub-objects), then commentAfter, *then* the final newline.

I'm still unsure of whether there should be a distinct KeyEntry type, and how to handle the naming. The current prototype uses `JsonDocument` and `CommentDocumentEntry` but these are cumbersome, and honestly if we wind up parsing and serializing in different formats like hocon, that language will become wrong.

**Reply by falkreon** [reply]
This should improve the readability of any accesses too, because you could potentially grab an iterator for only the keyValueEntries in an object, **which would include any associated comment information** based on our comment association rules, and not touch any of the user fluff around or in between the important parts. It would also help the internal implementation avoid scanning too far, and help keep validation of a document simple.

**Reply by falkreon** [reply]
continuing to tweak as I implement - if I hardcode a little bit more of the KeyValueDocumentEntry, such as:
```json5
keyValueEntry {
    key: "status"
    value: (reference hard equal (==) to the "arrayEntry" below)
    commentBefore: commentEntry "thing\nother thing" //folded
    commentAfter: commentEntry "number chosen by a fair diceroll; guaranteed to be random"
    contents: [
      commentEntry "foo"
      arrayEntry: (...)
    ]
}
```

The serialization flow for this would be to write commentBefore, then the key, then all entries, then commentAfter, followed by the newline. I'm not convinced that we need the ability to serialize comments before the colon, and in this scenario any comments deserialized before the colon, would be reserialized after it (but still before the key). The colon is "part of the key" to me, and that's important to the readability of the file, it's like fixing indents.

Speaking of which, we probably want to detect the indent style of the whole resource and stash it in the Document. Just count up the leading tabs and spaces as we scan, compare them to the indent levels, and get heuristic measurements:
* average number of leading spaces per indent level for spaces indents
* average number of leading tabs per indent level for tabs indents
* number of leading spaces and tabs to select whether spaces or tabs are used
And use this to assemble a String representation of a single indent level which gets stashed on Document. In the future, perhaps we can measure some other heuristics like brace style and stash those too.

Thinking I actually do want newlines tagged in the document, so that an array like
```json5
[ 1, 2, 3, /* foo */ 4,
  5, 6, 7, /* bar */ 8 ]
```
can pretty much survive unmangled in a full de/serialize cycle. Line-end comments require a newline, so they'd imply one. But an empty line-end comment is not the same thing as a raw line break, so trailing `//` would survive too, which can be important for multi-line-end comments that make up a larger block.

**Comment by exaskye** [comment]


**Reply by falkreon** [reply]
I'm not convinced that adding an asterisk to the front of a comment will make it clear to a non-technical end-user what the comment is for. I think if they see a document formatted like that, they'll just assume that's how you do comments - and then be confused when their comment is completely clobbered.

Yes, introspecting into documents can be difficult. Yes, things would be much easier if we could, instead of dealing with comments everywhere, specify that comments, or a certain type of comments, only go before keys. If we can have a hot path for dealing with the common case (such as observing that the document *only* has comments before keys, therefore comments may be attributed cleanly to keys), we should make that easy. But this does not help us preserve comments stuffed in odd places.

**Reply by falkreon** [reply]
really the attribution problem is more like
```json5
/* a */
{
  /* b */
  foo /* c */ : /*d*/ bar //e
}
```

What you seem to be suggesting is that `b` should be the only reasonable attribution position (and `a` for the object), but should we really make `e` undetectable?

**Reply by exaskye** [reply]
> I'm not convinced that adding an asterisk to the front of a comment will make it clear to a non-technical end-user what the comment is for.

I feel like it's the responsibility of the program creating a file that makes use of doc comments that is expected to be used by people who don't know what a doc comment *is* to briefly explain the concept in a header comment.

> But this does not help us preserve comments stuffed in odd places.

No. It's a suggestion for making a certain kind of special comment easier to introspect, while still using the document approach.

> What you seem to be suggesting is that b should be the only reasonable attribution position (and a for the object), but should we really make e undetectable?

All of the comments would be *detectable*, but only in the document view as explicit comment nodes. Only a and b would be available as doc comments if the API user chose to use them, and if they started with `/**` rather than `/*`.

**Reply by falkreon** [reply]
mostly on board, but this is somewhat at odds with the pillar of embracing the weirdness humans create in their files. Why does it *have* to be a doc comment to be detected?

**Comment by falkreon** [comment]
Starting to implement the Document view in mainline. How should we relate the semantic objects to the structural ones? What level of caching or defensive copying should occur? Does, say, a structural change in the document affect any derived semantic objects, and vise versa?
